SRC-2 Inline Documentation (#11)
FuelLabs/sway-standards: standards/src_2/README.md

<p align="center">
    <picture>
        <source media="(prefers-color-scheme: dark)" srcset=".docs/src-2-logo-dark-theme.png">
        <img alt="SRC-5 logo" width="400px" src=".docs/src-2-logo-light-theme.png">
    </picture>
</p>

> # Abstract

Reviewer: That's a lot of `h1` tags, perhaps we should alter this?

> The following standard intends to define the structure and organization of inline documentation for functions, structs, enums, storage, configurables, and more within the Sway Language. This will be a living standard.

Reviewer: ```suggestion
The following standard intends to define the structure and organization of inline documentation for functions, structs, enums, storage, configurables, and more within the Sway Language. This is a living standard.
```

# Motivation

> The standard seeks to provide a better developer experience using Fuel's tooling.

Reviewer: This sentence does not convey the context of what will be better for the user and how it relates to our tools. Could you provide some more context for the reader?

# Prior Art

> A number of pre-existing functions in the [sway standard library](https://github.com/FuelLabs/sway/tree/master/sway-lib-std), [sway-applications](https://github.com/FuelLabs/sway-applications), and [sway-libs](https://github.com/FuelLabs/sway-libs) repositories have inline documentation. The inline documentation for these is already compatible with Fuel's VS Code extension. These however do not all follow the same structure and outline.

Reviewer: I would remove the "prior art" section and merge it with the "motivation".
I'd probably remove any mention of the repositories and outline the benefit of a standard - how not all devs structure code in the same way, and thus the additional benefit of enabling the creation of tooling

Author: This follows the structure of standards as mentioned before.

# Specification

## Functions

The following describes the structure and order of inline documentation for functions. Some sections MAY NOT apply to each function. When a section is not relevant it SHALL be omitted. 

#### - Description

This section has no header. 
A simple explanation of the function's intent or functionality. 

> Example:
> ```rust

Reviewer: ```suggestion
Example:

```rust
```

Could you add new lines around the code blocks throughout the file? A long time ago John mentioned something about tooling benefiting from this

```

#### - Additional Information

This section has a `h3` header.
This section is directly below the description and can provide additional information beyond the enum's purpose or functionality.
Example:
```rust
/// ### Additional Information
///
/// This enum also has some complex behaviors.
```

### - Variant 

The following describes the structure and order of inline documentation for fields within enums. Some sections MAY NOT apply to each field. When a section is not relevant it SHALL be omitted. 

#### Description

This section has no header. 
Each variant SHALL have its own description with a simple explanation of the variant's purpose or functionality. 
Example:
```rust
/// This variant represents the uninitialized state of a contract.
variant_1: (),
/// This variant represents the initialized state of a contract.
variant_2: Identity,
```

#### - Additional Information

This section has a `h3` header.
This section is directly below the description and can provide additional information beyond the variant's purpose or functionality.
Example:
```rust
/// ### Additional Information
///
/// This variant also has some complex behaviors.
```

## Errors

> In Sway, errors are recommended to be enums. They SHALL follow the same structure and order for inline documentation as described above for enums. Some sections MAY NOT apply to each error. When a section is not relevant it SHALL be omitted. 

Reviewer: Again, I would move the examples from the bottom to their relevant sections.

Author: Each section already has an example and the examples section follows the structure of standards.

#### - Description

This section has no header. 
A simple explanation of the configurable variable's purpose or functionality. 
Example:
```rust
/// This configurable variable is used for an address.
```

#### - Additional Information

This section has a `h3` header.
This section is directly below the description and can provide additional information beyond the configurable variable's purpose or functionality.
Example:
```rust
/// ### Additional Information
///
/// This configurable variable makes security assumptions.
```

# Rationale

The SRC-2 standard should help provide developers with an easy way to both quickly write inline documentation and get up to speed on other developers' code. This standard in combination with Fuel's VS Code extension provides for readily accessible information on functions, structs, and enums

![Screenshot 2023-05-10 125656](https://github.com/FuelLabs/sway-standards/assets/54727135/f03073b9-2a28-44d1-b12a-5603a0738fee)

# Backwards Compatibility

There are no standards that the SRC-2 standard requires to be backward compatible with.

# Security Considerations

> This standard will improve security by providing developers with relevant information such as revert cases.

Reviewer: These should really be above the spec and I'd combine the rationale into the motivation as 1 section.

Reviewer 2: Security considerations are typically at the bottom because they require the context of understanding the proposed standard.

It's a little different for this one, since it's just a documentation standard, but it's still a good convention.

Reviewer: Fair point, having an understanding of the content is a pre-requisite to appreciating the security. At the same time you may argue that being aware of the security will aid in the understanding of the decisions behind the specification requirements and thus they should be listed somewhere towards the top.

Realistically, you'd have two documents. One would be an overview for a less technical person and another would be in-depth for someone like an engineer. The overview would likely contain security towards the top while the detailed one would probably have it at the end.

Author: Standards are inherently technical as they are used by engineers. Splitting a standard into 2 documents would only contribute to the dissection of information and make it more difficult to follow.

# Motivation

The standard seeks to provide a better developer experience using Fuel's tooling and the Language Server. This will allow for better interoperability between applications and enable developers to quickly understand any external code they are using or implementing.

# Prior Art

A number of pre-existing functions in the [sway standard library](https://github.com/FuelLabs/sway/tree/master/sway-lib-std), [sway-applications](https://github.com/FuelLabs/sway-applications), and [sway-libs](https://github.com/FuelLabs/sway-libs) repositories have inline documentation. The inline documentation for these is already compatible with Fuel's VS Code extension. These however do not all follow the same structure and outline.

# Specification

## Functions

The following describes the structure and order of inline documentation for functions. Some sections MAY NOT apply to each function. When a section is not relevant it SHALL be omitted. 

#### - Description

This section has no header. 
A simple explanation of the function's intent or functionality. 
Example:

```rust
/// This function computes the hash of two numbers.
```

#### - Additional Information

> This section has a `h3` header.

Reviewer: Purely for visual purposes

> This section has a `h3` header.

Reviewer 3: If that's the only reason I think we should most definitely produce properly formatted markdown and start titles at h1.

Formatting belongs in CSS.

> This section has a `h3` header.

Reviewer 4: @IGI-111 While I do agree with you in principle, we read and render this markdown directly to display formatted markdown in hover boxes in LSP. H1 tags are definitely too big in this instance and H3 is more readable. 

That said we also read and render this in forc-doc which does have it's own css where heading sizes can be tweaked.

> This section has a `h3` header.

Reviewer 4: It seems rust is using H1's in their doc comments and also rendering them in the hover pop up boxes at that size. I'm happy to follow their lead.
